binary_new_name does not expand template variables

## Problem

`with.binary_new_name` in `action.yml` is documented as "string", but unlike `download_url`, `tarball_binary_path`, and `smoke_test`, it does **not** undergo template substitution for `\${binary}` or `\${version}`.

`index.js` calls `fillTemplate` only on `downloadURL`, `tarballBinaryPath`, and `smokeTest`. `binaryNewName` is read from inputs and passed directly into `tar --transform=s/.../<value>/`.

As a result, callers who pass `binary_new_name: "\${binary}"` (or any string containing template vars) get the literal text `\${binary}` as the renamed binary. Subsequent steps that rely on the binary being on PATH then fail with:

\`\`\`
##[error]Unable to locate executable file: <name>
\`\`\`

This was the root cause of pre-commit failures across multiple giantswarm repos in the 2026-04-20 marge sweep (klaus, backstage, klaus-operator, kserve, llm-testing, mcp-kubernetes, mcp-prometheus). The immediate fix in giantswarm/github#5082 hardcoded the binary name to work around the missing substitution.

## Acceptance criteria

- `binary_new_name` accepts and expands the same template variables as the other inputs (`\${binary}`, `\${version}`).
- Document the supported substitutions for each input in the README so the contract is explicit.
- Add a smoke test that exercises a templated `binary_new_name`.

## Workaround

Pass a literal binary name in `binary_new_name`. This is what giantswarm/github#5082 does for `helm-schema`.

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -98,10 +98,7 @@ jobs:
       # End-to-end coverage of the install steps used in the central
       # pre-commit workflow at giantswarm/github/workflows/zz_generated.pre-commit.yaml.
       # These tarballs ship the binary at the archive root (no subdirectory),
-      # so tarball_binary_path must not contain a wildcard/slash. The
-      # helm-values-schema-json archive additionally names the binary
-      # "schema", so binary_new_name must be supplied as a literal (inputs
-      # other than download_url/tarball_binary_path are NOT template-expanded).
+      # so tarball_binary_path must not contain a wildcard/slash.
       - name: Test install of helm-docs (flat tarball, no rename)
         uses: ./
         with:
@@ -114,14 +111,17 @@ jobs:
         run: |
           which helm-docs
           helm-docs --version
-      - name: Test install of helm-values-schema-json (flat tarball, rename schema -> helm-schema)
+      # Regression test for https://github.com/giantswarm/install-binary-action/issues/339:
+      # binary_new_name must support the same ${binary}/${version} template
+      # substitution as download_url, tarball_binary_path, and smoke_test.
+      - name: Test install of helm-values-schema-json (templated binary_new_name)
         uses: ./
         with:
           binary: helm-schema
           version: "2.3.1"
           download_url: "https://github.com/losisin/helm-values-schema-json/releases/download/v${version}/helm-values-schema-json_${version}_linux_amd64.tgz"
           tarball_binary_path: "schema"
-          binary_new_name: "helm-schema"
+          binary_new_name: "${binary}"
           smoke_test: "${binary} version"
       - name: Verify helm-schema is on PATH
         run: |

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Apply `${binary}` / `${version}` template substitution to `binary_new_name`,
+  matching the behaviour already documented for `download_url`,
+  `tarball_binary_path`, and `smoke_test`
+  ([#339](https://github.com/giantswarm/install-binary-action/issues/339)).
+
 ## [4.0.1] - 2026-04-17
 
 ### Fixed

--- a/README.md
+++ b/README.md
@@ -32,7 +32,9 @@ template format. Templated variables are `${binary}` and `${version}`.
 
 ## `binary_new_name`
 
-**Optional.** Name of the binary once is uncompressed and installed in the selected directory.
+**Optional.** Name of the binary once is uncompressed and installed in the
+selected directory. Can be in template format. Templated variables are
+`${binary}` and `${version}`.
 
 ## Outputs
 

--- a/action.yml
+++ b/action.yml
@@ -20,7 +20,7 @@ inputs:
     required: false
     default: '${binary} version'
   binary_new_name:
-    description: 'Name of the binary once is installed. E.g. "aws"'
+    description: 'Name of the binary once is installed. E.g. "aws". Available variables are ${binary} and ${version}.'
     required: false
 outputs: {}
 runs:

--- a/index.js
+++ b/index.js
@@ -10,7 +10,7 @@ const run = async () => {
   try {
     const binary = core.getInput('binary');
     const version = core.getInput('version');
-    const binaryNewName = core.getInput('binary_new_name');
+    let binaryNewName = core.getInput('binary_new_name');
     let downloadURL = core.getInput('download_url');
     let tarballBinaryPath = core.getInput('tarball_binary_path');
     let smokeTest = core.getInput('smoke_test');
@@ -23,6 +23,7 @@ const run = async () => {
     downloadURL = fillTemplate(downloadURL);
     tarballBinaryPath = fillTemplate(tarballBinaryPath)
     smokeTest = fillTemplate(smokeTest);
+    binaryNewName = fillTemplate(binaryNewName);
 
     core.info(`binary:               ${binary}`);
     if (binaryNewName) {

